the-collab-lab/tcl-46-smart-shopping-list
# Sorted list by purchase priority and name

## Description
This update passes the data from firebase through a sorting function so that all list items are prioritized by purchase urgency on the List page. These are ordered in the following groups from top (highest priority) to bottom:
1. **Overdue** (dateNextPurchased has passed, but item is _active_: fewer than 60 days have passed since the last purchase or dateCreated)
2. Need to buy **soon** (7 days or fewer until the next purchase; active)
3. Need to buy **kind of soon** (between 7 & 30 days until the next purchase; active)
4. Need to buy **not soon** (30 days or more until the next purchase; active)
5. **Inactive** (60 days have passed since the last purchase)

Items in each group are in ascending order of days to next purchase.

Where the estimated days to next purchase are equal, the items are sorted by name alphabetically.

UI in the form of both plain text (bolded above) and color indicators are added to each ListItem so the user knows how their items are sorted. A placeholder for [Issue 11](https://github.com/the-collab-lab/tcl-46-smart-shopping-list/issues/11) delete button is also included. All final styling TBD with the team.

### Considerations
- Despite the acceptance criteria below, `comparePurchaseUrgency` in the current approach does not interact with the Firebase API and seems more suited as a utility function. It is currently inside `item.js` in .utils.
- To run the sorting operations more efficiently inside an overall `comparePurchaseUrgency` function, we attached keys to each item (daysToNext, refTime). In the latest update, these keys are also passed to and used in `ListItem` for urgency labeling.
- We decided to consider items that have never been purchased (no "last purchase date") as inactive if 60 days have passed since dateCreated, which also aligns with the approach for [Issue 10, Estimate purchase date](https://github.com/the-collab-lab/tcl-46-smart-shopping-list/issues/10).
- The current approach makes no additional requests to the server for sorting after the initial data request.
- Current alphabetizing/lexicographic sorting uses localeCompare, which may treat special characters variably depending on locale.

## Related Issue

Closes #12 

## Acceptance Criteria
- [x] `api/firestore.js` exports a new `comparePurchaseUrgency` function with the following behaviors (see above note)

## Type of Changes



|     | Type                       |
| --- | -------------------------- |
|     | :bug: Bug fix              |
| ✓   | :sparkles: New feature     |
|     | :hammer: Refactoring       |
|     | :100: Add tests            |
|     | :link: Update dependencies |
|     | :scroll: Docs              |

## Updates

### Before

<img width="700" alt="before" src="https://user-images.githubusercontent.com/56039221/184168359-88235f09-ad56-468a-876f-b63131f089c9.png">


### After

<img width="700" alt="after" src="https://user-images.githubusercontent.com/56039221/184168382-86fd3320-693e-42a4-a399-734362bf6056.png">

<br>

## Notes on UI Changes

The different urgency categories are indicated in two separate ways:

- **Text** - the "Soon", "Kind Of Soon", etc

- **Other** - Something like the different color dots. Other ideas include giving each list item a background color, or an SVG icon, that correspond to the different urgency categories

To accomplish this, each `ListItem` includes a string, `urgencyMessage`, wrapped inside of a `<p>` element

The color dot is a `<div>` with a `className` of `{urgency}`. The `{urgency}` can be set to `soon`, `kind-of-soon`, `overdue`, and other names that correspond to the list item's urgency category. 

This way, we can change the "Other" indicator based off of the `{urgency}` of each item by adding styles to each of those classes

<br>

**To make these changes visible, some temporary styles were added:**

- Background color and spacing for `ListItem` components 
- Put each sub-element (checkbox, label, urgency, delete) of `ListItem` into it's own `<div>` just to make quick adjustments easier. Not the best practice so this should only be temporary
- A placeholder for the `delete` button
- Each `<div>` has it's width set as a percentage of the entire `ListItem` component. (currently 10% / 40% / 30% / 20%)

<img width="700" alt="flexbox" src="https://user-images.githubusercontent.com/56039221/184187901-2a99de63-a65b-408a-8e09-55c274d0b011.png">

The goal here was to add enough to the UI to make the urgency indicator UI updates make sense, but avoid spending too much time on UI during this week's issue. We should decide and improve on the UI / styling as a team in the final weeks




## Testing Steps / QA Criteria

1. Pull and checkout the `ad-hy-sortedList-12-test` branch. Run with `npm start`.

2. Generate a new list and add some items with different purchase frequency; the List view should reflect the right order of priority based on the estimated next purchase date, and then alphabetically by name.
3. When you check off an item, note that the item position will shift where applicable (generally with "Not soon" items compared to other starting frequencies, if you click twice - with 1 min elapsed between - you will see a more dramatic shift based on the calculateEstimate algorithm).
4. To see items categorized "Inactive", find the list in the Firebase console and manually edit one item: either dateCreated (if !dateLastPurchased) or dateLastPurchased and dateCreated* to a date 60 or more days before current date of access. The Inactive item should move to the bottom of the list. _*To avoid visual contradiction, you're advised to uncheck any items you are manually updating this way in Firebase console. Similar recommendation applies to not assigning a dateLastPurchased that's earlier than dateCreated for the same item._
5. To see "Overdue" items, in the Firebase console, edit the dateNextPurchased of an item to a date prior to current date of access. Note that the item will need to not be categorized inactive for the label to appear (otherwise, "inactive" takes precedence). The Overdue item should also move to the top of the list.
7. Additional tests (on checkIfActive and sortByNextDateAlphabetical functions) can be run with `npm test`.
